datadotworld/filewriter.py

```diff
@@ -0,0 +1,166 @@
+        """
+        write the given value to the stream - if the object is a bytearray,
+        write it as-is - otherwise, convert the object to a string with
+        `str()` and write the UTF-8 bytes
+
+        Parameters
+        ----------
+        value: str or bytearray
+            the value to write
+        """
+        if 'w' == self._mode:
+            if isinstance(value, str):
+                self._queue.put(value.encode('utf-8'))
+            else:
+                raise TypeError(
+                    "write() argument must be str, not {}"
+                        .format(type(value)))
+        elif 'wb' == self._mode:
+            if (isinstance(value, (bytes, bytearray))):
+                self._queue.put(value)
+            else:
+                raise TypeError(
+                    "a bytes-like object is required, not {}"
+                        .format(type(value)))
+        else:
+            raise NotImplementedError(
+                "modes other than 'w' and 'wb' not supported")
+
+    def open(self):
+        """
+        start the thread executing the HTTP request
+        """
+        def put_request(body, response_queue, host,
+                        config, dataset_key, file_name):
+            ownerid, datasetid = parse_dataset_key(dataset_key)
+            response = requests.put(
+                "{}/uploads/{}/{}/files/{}".format(
+                    host, ownerid, datasetid, file_name),
+                data=body,
+                headers={
+                    'Authorization': 'Bearer {}'.format(config.auth_token)
+                })
+            response_queue.put(response)
+
+        body = iter(self._queue.get, self._sentinel)
+        self._thread = Thread(target=put_request,
+                              args=(body,
+                                    self._response_queue,
+                                    self._api_host,
+                                    self._config,
+                                    self._dataset_key,
+                                    self._file_name))
+        self._thread.start()
+
+    def close(self):
+        """
+        closing the writer adds the sentinel value into the queue and joins
+        the thread executing the HTTP request
+        """
+        self._queue.put(self._sentinel)
+        self._thread.join(timeout=self._timeout)
+        if self._thread.is_alive():
+            raise RemoteFileException("Closing file timed out.")
+        response = self._response_queue.get_nowait()
+        try:
+            response.raise_for_status()
+        except Exception as e:
+            raise RemoteFileException(cause=e)
+
+    def __enter__(self):
+        """
+        Implement the context manager protocol
+        """
+        self.open()
+        return self
+
+    def __exit__(self, e_type, e_value, traceback):
+        """
+        Implement the context manager protocol
+        """
+        self.close()
+
+
+class RemoteFileException(Exception):
```

Comment:
Consider renaming to `RemoteFileError`. `Error` is a more common suffix in Python.